packages/grid/_modules_/grid/hooks/features/rowGrouping/useGridRowGrouping.tsx

```diff
@@ -0,0 +1,475 @@
+        row,
+        id,
+        colDef,
+      }: {
+        row: GridRowModel;
+        id: GridRowId;
+        colDef: GridColDef;
+      }) => {
+        let value: GridCellValue;
+        if (colDef.valueGetter) {
+          const valueGetterParams: GridValueGetterSimpleParams = {
+            colDef,
+            field: colDef.field,
+            id,
+            row,
+            rowNode: {
+              isAutoGenerated: false,
+              id,
+            },
+          };
+          value = colDef.valueGetter(valueGetterParams);
+        } else {
+          value = row[colDef.field];
+        }
+
+        let key: GridKeyValue | null | undefined;
+        if (colDef.keyGetter) {
+          key = colDef.keyGetter({
+            value,
+            id,
+            field: colDef.field,
+          });
+        } else {
+          key = value as GridKeyValue;
+        }
+
+        return {
+          key,
+          field: colDef.field,
+        };
+      };
+
+      params.ids.forEach((rowId) => {
+        const row = params.idRowsLookup[rowId];
+
+        rowGroupingModel.forEach((groupingCriteria) => {
+          const { key } = getCellGroupingCriteria({
+            row,
+            id: rowId,
+            colDef: columnsLookup[groupingCriteria],
+          });
+          const groupingFieldsDistinctKeys = distinctValues[groupingCriteria];
+
+          if (key != null && !groupingFieldsDistinctKeys.lookup[key.toString()]) {
+            groupingFieldsDistinctKeys.lookup[key.toString()] = true;
+            groupingFieldsDistinctKeys.list.push(key);
+          }
+        });
+      });
+
+      const rows = params.ids.map((rowId) => {
+        const row = params.idRowsLookup[rowId];
+        const parentPath = rowGroupingModel
+          .map((groupingField) =>
+            getCellGroupingCriteria({
+              row,
+              id: rowId,
+              colDef: columnsLookup[groupingField],
+            }),
+          )
+          .filter((cell) => cell.key != null) as BuildRowTreeGroupingCriteria[];
+
+        const leafGroupingCriteria: BuildRowTreeGroupingCriteria = {
+          key: rowId.toString(),
+          field: null,
+        };
+
+        return {
+          path: [...parentPath, leafGroupingCriteria],
+          id: rowId,
+        };
+      });
+
+      return buildRowTree({
+        ...params,
+        rows,
+        defaultGroupingExpansionDepth: props.defaultGroupingExpansionDepth,
+        groupingName: GROUPING_COLUMNS_FEATURE_NAME,
+      });
+    };
+
+    return apiRef.current.unstable_registerRowGroupsBuilder('rowGrouping', groupRows);
+  }, [apiRef, props.defaultGroupingExpansionDepth, props.disableRowGrouping]);
+
+  useFirstRender(() => {
+    updateRowGrouping();
+  });
+
+  const isFirstRender = React.useRef(true);
+  React.useEffect(() => {
+    if (isFirstRender.current) {
+      isFirstRender.current = false;
+      return;
+    }
+
+    updateRowGrouping();
```

Comment:
I noticed the same in `useTreeData`.

```suggestion
    return updateRowGrouping();
```

Reply:
What is the advantage of the return here ?
There is no un-subscribe behavior on the row grouping.
If I add the `return`, eslint complains that the return are inconsistent.
I could return `() => {}` inside the `if` above or disable the lint rule for this line but I don't think it's interesting to do it.